### ISSUE TITLE
More generalizable formatting expressions?

### ISSUE THREAD

**Issue by ErichDonGubler** [Lucretiel/lazy_format#2]
`if let` would be a nice addition to what the current macro handles:

```rust
let display_response = lazy_format!(
    "{}{}\n...with response body: {}",
    status_code.as_str(),
    lazy_format!(
        if let Some(r) = status_code.canonical_reason() {
            (" (\"{}\")", r)
        } else {
            ("")
        }
    ),
    body,
);
```

The only catch is that this adds a second possibility for the format of the `lazy_format!` macro -- which might be complex if naively implemented. I'm not sure if there IS a clean way to do with declarative macros. One could also solve the problem like this:

```rust
let display_response = lazy_format!(
    "{}{}\n...with response body: {}",
    status_code.as_str(),
    lazy_format!(
        if let Some(r) = status_code.canonical_reason() {
            (" (\"{}\")", r)
        } else {
            ("")
        }
    ),
    body,
);
```

...but this use case is actually a strawman to propose something else. ;)

Another way to implement the above use case might be to expose a library struct very similar to the one created by `LazyFormat` (sample strawman name: `DisplayWith`) but with two changes:

* Don't add a `Debug` impl, since that depends on the expression
* Make the newtype member of `DisplayWith` `pub`, so that client code can write code of the form:

    ```rust
    let display_response = DisplayWith(move |f| {
        write!(f, "{}", status_code.as_str())?;
        if let Some(r) = status_code.canonical_reason() {
            write!(f, " (\"{}\")", r)?;
        }
        write!(f, "\n...with response body: {}", body)
    });
    ```

I've actually used code exactly like this with this implementation, which you may feel free to use:

```rust
use std::fmt::{Debug, Display, Formatter, Result as FmtResult};

pub struct DisplayWith<F: Fn(&mut Formatter) -> FmtResult>(pub F);

impl<F: Fn(&mut Formatter) -> FmtResult> Debug for DisplayWith<F> {
    fn fmt(&self, f: &mut Formatter) -> FmtResult {
        <Self as Display>::fmt(self, f)
    }
}

impl<F: Fn(&mut Formatter) -> FmtResult> Display for DisplayWith<F> {
    fn fmt(&self, f: &mut Formatter) -> FmtResult {
        (self.0)(f)
    }
}
```

HOWEVER, a ***third*** idea might be to tweak the syntax slightly and allow lambdas to be specified by the `lazy_format!` macro, 

```rust
let display_response = lazy_format!(move |f| { // `move` keyword is optional
    write!(f, "{}", status_code.as_str())?;
    if let Some(r) = status_code.canonical_reason() {
        write!(f, " (\"{}\")", r)?;
    }
    write!(f, "\n...with response body: {}", body)
});
```

I would actually be in favor of this latter solution, since it would allow arbitrarily-complex expressions but keep the clean macro usage with a `Debug` impl. Then again, solutions 2 and 3 are orthogonal solutions, and 2 might be desirable for those who want less "magic" in their code.

If you'd prefer me to submit a PR with my suggestions, I'm more than happy to do so! :)

**Comment by ErichDonGubler** [comment]
Ah, derp, I've just noticed `make_lazy_format`, which IS the third option.

Seems like my particular use cases are solved, then. I did this, for future reference of anyone who was interested in this issue after closing:

```rust
let display_response = lazy_format!(
    "{}{}\n...with response body: {}",
    status_code.as_str(),
    lazy_format!(
        if let Some(r) = status_code.canonical_reason() {
            (" (\"{}\")", r)
        } else {
            ("")
        }
    ),
    body,
);
```

**Comment by Lucretiel** [comment]
I'll see about adding `if let`, since I already support ordinary `if`

**Comment by Lucretiel** [comment]
Fixed in [1.6.0](https://github.com/Lucretiel/lazy_format/tree/1.6.0). Thanks for the idea!

Caveats:

- Doesn't support guards or if-let chaining in `if-let`-style formatting
- Added support for pattern alternation in `|`, both in `if` and `match` formatting

**Comment by Lucretiel** [comment]
Regarding your idea about a public struct: There actually is such a struct, flagged as `#[doc(hidden)]` I've hedged back-and-forth on whether to make it part of the public interface. Right now I'm leaning towards not doing that, because I want to keep the flexibility to freely change the implementation details of the struct, which I've already done a couple times throughout this project.

My design philosophy so far has been to try to add sensible, "common" use cases to `lazy_format!`, and allow `make_lazy_format!` to support all other use cases. In particular, I'm looking for cases like this:

```
// Doesn't work because the separate branches have different types
let result = if condition {
    lazy_format!(...);
} else {
    lazy_format!(...);
}
```

In this case, the different branch blocks have different types, so the code fails to compile. To support this, I added syntax so that the `if` could be present *inside* the `lazy_format`, so that the whole thing can have a consistent type.
